### PR TITLE
Dampen telefrag behavior for player in particular

### PR DESCRIPTION
**Pull request by anoobindisguise**
#### Summary
Balance "Don't telefrag the player in most cases"

#### Purpose of change
Telefrags-on-player suck ass. It's not a fun mechanic, it's just "did you not have an active anchor? here's a chance to instantly die in a single turn lol" which we've been moving away from - among other "!FUN!" elements such as tankbots, lava rifts, chicken walkers, flaming eyes, and so on.

#### Describe the solution
Any time an entity would be fragged, there is a 50% chance they will be teleported away instead of being fragged. If the player would be fragged, there is a 100% chance they will be teleported away unless they have teleglow, in which case they have the same 50% chance to frag or be teleported that every other entity has.

This makes playing with portals still dangerous and incentivizes using an anchor - a random teleport is really unsafe and if you don't take it as a warning to GTFO you run the risk of getting fragged. Therefore there is counterplay instead of just instantly dying no matter the character. This also preserves the lore element of telefrags - telefrags are established as things that can happen in multiple lore snippets, and the random chance element helps explain why the player isn't just magically immune by virtue of being the player - they're just getting lucky. But for the average player who doesn't dive code, it should be a convincing magic trick.

#### Describe alternatives you've considered
make the player entirely immune no matter what

#### Testing
Compiled successfully, spawned the tindalos lab finale and equipped power armor. Sat still a very long time until something would have fragged me, at which point I was teleported away instead of being fragged. Observed enemies occasionally getting fragged and occasionally getting teleported. Due to the extreme rarity of telefrags it's a difficult subject to test so I can't guarantee this is working 100% correctly. There is probably also a very very small chance that something that would get fragged by a teleporter gets teleported to the teleporter's space instead and counter-frags them before they can teleport but I have not observed this phenomenon and have no idea what would happen if it occurred. I don't want to make the teleport SAFE because there is a chance it could fail if it doesn't find a valid destination, instead it could be replaced with a function that loops safe teleports until it makes a successful one.  If someone feels this is necessary I can do that too (I think)

### CODE REVIEW

**Comment by anoobindisguise** [comment]
Is it worth adding the following?
`    critter.setpos( target );
    //player and npc exclusive teleporting effects
    if( p ) {
        g->place_player( p->pos() );
        if( add_teleglow ) {
            p->add_effect( effect_teleglow, 30_minutes );
        }
    }
    if( c_is_u ) {
        g->update_map( *p );
    }
    critter.remove_effect( effect_grabbed );`
    
`    if( critter ){
        critter.setpos( target );
        //player and npc exclusive teleporting effects
        if( p ) {
            g->place_player( p->pos() );
            if( add_teleglow ) {
                p->add_effect( effect_teleglow, 30_minutes );
            }
        }
        if( c_is_u ) {
            g->update_map( *p );
        }
        critter.remove_effect( effect_grabbed );
    }
    return true;`
   


It probably doesn't hurt in the event that somehow the recursive function causes the original `critter` to be deleted by dying.

**Comment by anoobindisguise** [comment]
I attempted to implement a sanity check but the method I used [if( creature )] doesn't work in c++, I'm used to Java lol. It's probably not worth worrying about.

**Comment by John-Candlebury** [comment]
I thematically dislike telefrag being something that can happen to monsters but not to players, it feels very artificial.

Some of my locations (the portal lab specifically) are designed under the assumption that telefragging is a real risk.

**Comment by anoobindisguise** [comment]
> I thematically dislike telefrag being something that can happen to monsters but not to players, it feels very artificial.
> 
> Some of my locations (the portal lab specifically) are designed under the assumption that telefragging is a real risk.

It is worth reiterating that the player is only immune if they currently don't have teleglow, and monsters exhibit the same teleport-away behavior some of the time giving the player one "grace telefrag" meaning if they press on in a dangerous place they can still get fragged. Whether or not the player should get a grace telefrag is up for debate but the player definitely isn't immune outright.

**Comment by RelMayers** [comment]
Are you aware how unlikely player is to be telefragged?
Also arent you robbing player of possible story?

**Comment by RelMayers** [comment]
Also your testing is bad.

**Comment by RelMayers** [comment]
XXXXX
XMTPX
XXXXX
Teleport into solid rock deep underground, remove 3 tiles of solid rock so it forms a line
Place a teleporter trap in Middle spawn monster on one side and be with player on other side. Can be adjusted for NPCs testing as well.

**Comment by anoobindisguise** [comment]
> Are you aware how unlikely player is to be telefragged? Please describe likelyhood of player being telefragged. Also arent you robbing player of possible story?

That's a really bad argument. An event being unlikely doesn't make it impossible, and with telefrag being instant death, any chance of actual instant death from a mundane action shouldn't be allowed. I've been fragged by something I couldn't see before, and not because I was in a distorted lab or research facility, but simply from walking around a microlab and getting zapped by a tear in reality I didn't see. (Candlebury suggested just removing that specific portal which is fair also but doesn't address the root problem)
Further the game specifically tries multiple times to telefrag something whenever something teleports. Given an open field with you as the only target, your odds of getting fragged by a teleport that features you in its radius is about 1/100. It can be higher or lower in practice, because if there are walls, the game tries 20 times to find a space that isn't a wall (instant death if it fails) so in an area with many walls the odds go up, and if there are many targets to find there's a chance of the "check 5 times for frag" loop short circuiting on someone else. So your odds of death whenever something teleports with you around are, at their absolute best, about 1/550. Those really aren't good odds when they compound over the course of many hours. And if you wander into the distorted microlab without the anchor and ignore the sign, fine, you probably deserve it. But there are a lot of random portals lying around.

Generally speaking being 1-shot doesn't make for a good story either. It's basically "rocks fall and everyone dies" as far as the gameplay is concerned.

**Comment by RelMayers** [comment]
Please reply to rest of arguments. Also why answer for bad teleport is another teleport?

**Comment by anoobindisguise** [comment]
> Please reply to rest of arguments. Also why answer for bad teleport is another teleport?
I don't want to completely negate the "lore" of teleports, because they are supposed to be dangerous and unpredictable. This isn't a 100% foolproof "you'll never get telefragged again" but it's an improvement that in my opinion doesn't mess up the current notion that you really do not want to teleport or exist near other things that are being teleported if you value your life.

> Also your testing is bad. "realistic" to have someone magically be teleported away???
Teleportation is already literally just magic. It can do whatever we want. It isn't bound by any physical rules, it does not need to work one particular way. The only precedent is that telefragging is a common "trope" in other media - it originated in DOOM IIRC. (Although in DOOM the player could only be telefragged in the last level of the game, and DOOM isn't a permadeath rogue-like-adjacent-RPG.)

> XXXXX
> XMTPX
> XXXXX
> Teleport into solid rock deep underground, remove 3 tiles of solid rock so it forms a line
> Place a teleporter trap in Middle spawn monster on one side and be with player on other side. Can be adjusted for NPCs testing as well.

Your hypothetical scenario never actually occurs at any point. If that happened, yes, whatever teleported would probably teleport into solid rock and die. But solid objects are usually at most 50% of available teleport spaces, and under those circumstances a teleporter have a 0.5^20 chance to cause that.

**Comment by I-am-Erk** [comment]
@RelMayers - if you can't frame your feedback in a constructive way, find another outlet for your crankiness. "Also your testing is bad." is useless feedback, you are wasting our time.

**Comment by I-am-Erk** [comment]
@anoobindisguise Bear in mind that the tile that a character occupies is not 100% full of character. The likelihood of two things occupying exactly the same space even in one tile is very very low. I think a better mechanism for this whole thing would be that if two things are teleported into the same space, they both take random damage that isn't reduced by armour, inflicted to limbs, and are knocked away 1-2 tiles and prone. Basically there's a small compression burst when they overlap, but we assume it's not their whole bodies that are overlapping.

**Comment by anoobindisguise** [comment]
> @anoobindisguise Bear in mind that the tile that a character occupies is not 100% full of character. The likelihood of two things occupying exactly the same space even in one tile is very very low. I think a better mechanism for this whole thing would be that if two things are teleported into the same space, they both take random damage that isn't reduced by armour, inflicted to limbs, and are knocked away 1-2 tiles and prone. Basically there's a small compression burst when they overlap, but we assume it's not their whole bodies that are overlapping.

OK, I'll take a look at that and see if I can get it implemented.

**Review by kevingranade — CHANGES_REQUESTED** [review]
Don't use teleporting, use the fling code it's already set up to handle all of these special cases and a dozen lines of new code is going to be riddled with bugs because it's full of edge cases.
Ideally fling both the teleportee and the target in opposite directions.

File: src/teleport.cpp [hunk 1]
```diff
@@ -168,7 +168,7 @@ bool teleport::teleport_to_point( Creature &critter, tripoint target, bool safe,
     }
     critter.setpos( target );
     if( collision ) {
-        g->fling_creature( critter, units::from_degrees( collision_angle - 180 ), 50 );
+        g->fling_creature( get_creature_tracker().creature_at<Creature>( target ), units::from_degrees( collision_angle - 180 ), 50 );
```
Comment (Stadler76) [inline src/teleport.cpp:171]:
```suggestion
        &critter, units::from_degrees( collision_angle - 180 ), 50 );
```
More or less a wild guess from doing a find in files 'fling_creature', but does this do the trick?

**Review by Stadler76 — COMMENTED** [review]
This might do the trick ... I guess?

**Comment by anoobindisguise** [comment]
OK, tested under the new paradigm, the double knock-away is a little janky but it's functional. observed both myself and enemies getting flung away. One time I was surrounded by mobs on all sides, and when this happened I took damage 5 times and then the teleportee flickered and didn't teleport. Presumably if the player would get thrown in a direction that causes them to not move any squares you instead take damage and try again up to 5 times, in which case it fails. This didn't oneshot me from full health, but it did a ton of damage. The damage pierces debug invincibility even. But it appears to be working now. Could be good to spawn an explosion SFX at the location, I'll look into that.

**Comment by Stadler76** [comment]
> Only remaining issue is that the fling chooses a random direction so it may repeat multiple times. Should the poor_soul be teleported if they're still in the same spot after being flung?

Maybe fling it one tile further but in a random direction, if flinging in the opposite direction was unsuccessful. This should handle cases, where the poor_soul is standing in a corner or in a dead end.

**Comment by Stadler76** [comment]
> > Only remaining issue is that the fling chooses a random direction so it may repeat multiple times. Should the poor_soul be teleported if they're still in the same spot after being flung?
> 
> Maybe fling it one tile further but in a random direction, if flinging in the opposite direction was unsuccessful. This should handle cases, where the poor_soul is standing in a corner or in a dead end.

Replying to myself here: The idea was, that if the poor_soul is standing against a wall, handle it as if it bounced off said wall. If it still didn't move, then don't try to fling and just check, if it's still sharing a space with the teleportee. If true, then telefrag, otherwise we're done. No additional damage to the poor_soul and call it a success.

**Comment by anoobindisguise** [comment]
Solution: loop breaks if collision= true. If you are flung in a wall, the teleportee is flung in the opposite direction, so unless you are in a tiny narrow hallway it will be knocked away from you. Otherwise you'll occupy the same space for a turn, but that's such a minor edge case I'm not gonna worry about it.

**Comment by I-am-Erk** [comment]
Still desired, appears mostly ready. Added freeze label but if @kevingranade wants to take a look and see if this meets requirements we could slide it in.

**Comment by anoobindisguise** [comment]
It's been awhile so I'll recompile this and make sure it's still working, will update later today.

**Comment by anoobindisguise** [comment]
Yep, it does seem to be working, I haven't seen any errors in testing. Up to Kevin if it's still too jank though.

**Pull request merged**